### ISSUE TITLE
how to 'signal' load

### ISSUE THREAD

**Issue by cekvenich** [muicss/johnnydepp#1]
I like this new one. In old one I had code like:
      link(href= ROOT + 'assets/css/spectre.css', onload='cssLoaded()', rel='stylesheet')
and in cssLoad() I could signal that css is loaded. 

I think now I can't do that.

**Comment by cekvenich** [comment]
or function onDeviceReady() { // nothing will work before this
   console.log('deviceready!')
   loadjs.done('device')
}

Called in phonegap.

**Comment by amorey** [comment]
The `require` method accepts a success callback that will be called after the dependencies have been met. The library will trigger a one-time-only download the first time it is used by `require`:
```javascript
depp.require(['plugin1', 'plugin2'], function() {
  /* all files in bundles 'plugin1' and 'plugin2' have been loaded */
});
```

**Comment by cekvenich** [comment]
So it works and I have convert 3 projects to it.
But it only works for the 3 file types. What if I have to do something else?
eg: calculate prime to 10 digits - then do something.

So I have to be able to manually say something required has been done. I can do that in loadjs. I need this feature.

**Comment by amorey** [comment]
It sounds like you're looking for generic pub/sub functionality so you can do things like:
```javascript
subscribe('my event', function() {
  // do something after calculating prime to 10 digits
});

depp.require(['plugin1', 'plugin2'], function() {
  // calculate prime to 10 digits
  publish('my-event');
});
```
Am I understanding the problem correctly?

**Comment by wgehner** [comment]
```
x = 5; //do something, then
depp.publish(''xxdone")

depp.require(['plugin1','plugin2','xxdone'], function(){
   console.log(x); //should print 5
})
```
Make sense? During page load, the 'publish' may come first and is likely executed before the require, but the require needs to know that the 'publish' happened. So it's more like a non-consuming stream than a classic publish/subscribe.

**Comment by amorey** [comment]
If the publish comes first then it can be implemented with `.define()`:
```javascript
var x = 5;  // do something, then
depp.define({'xxdone': []});

depp.require(['plugin1', 'plugin2', 'xxdone'], function() {
  console.log(x);  // should print 5
});
```

However, handling the case where the publish can come after complicates the nested loading logic significantly. Consider this scenario:

```javascript
depp.define({
  'plugin1': ['/file1.js'],
  'plugin2': ['/file2.js'],
  'plugin3': ['xxdone', '/file3.js']
});

depp.require(['plugin1', 'plugin2', 'xxdone'], function() {
  console.log(x);
});

depp.require(['plugin3'], function() {
  // "xxdone" will be treated as file which will cause load error
});

var x = 5;
depp.done('xxdone');
```

If "xxdone" is not pre-defined as a bundle then the library will treat it as a file when it tries to load "plugin3".

Will a separate pub/sub library solve your problem?

**Comment by wgehner** [comment]
Since `depp.require(['plugin3']` could be expressed as `depp.require(['xxdone','file3plugin']`, that wouldn't be a killer. But the following cases would need to work:

```
depp.require(['plugin1','plugin2','xxdone'], function(){
   console.log(x); //should print 5
})

x = 5; //do something, then
depp.define({'xxdone':[]})
```
And:

```
depp.require(['plugin1','plugin2','xxdone'], function(){
   console.log(x); //should print 5
 y =x +1;
  depp.define({'yydone':[]})
})

depp.require(['yydone'], function(){
 console.log(y); //should print 6
}
```
Loadjs does the loosely coupled signaling of a (simulated) load complete perfectly with `loadjs.done('xxdone')`. Haven't seen any others that do it this elegantly. The only quirk of loadjs was that that bundle ID comes after the library list, which is addressed by depp. Adding depp.done would take depp over the top. Using a different library presumably wouldn't allow me to use depp.require(['plugin1', 'plugin2', 'xxdone']. Would adding `depp.done(name)` as a shortcut for `depp.define({name:[]})` be a possibility?

**Comment by amorey** [comment]
One way to get around the bundle definition issue is to use `null` as a placeholder:
```javascript
depp.define({
  'plugin1': ['/file1.js'],
  'plugin2': ['/file3.js'],
  'plugin3': ['xxdone', '/file3.js'],
  'xxdone': null
});
```

This will make the nested loading logic simpler. However, there's another issue with loading order:
```javascript
depp.require(['plugin1', 'xxdone', 'plugin2'], function() {
  console.log(x);
});
```

Normally, JohnnyDepp loads files in parallel and executes them in series but if "xxdone" appears before "plugin2" then to guarantee execution order the library will have to wait to load "plugin2" until after `deep.done('xxdone')` is called. The same is true of "plugin3". This behavior violates one of the key promises of JohnnyDepp - load files in parallel and execute them in series. Incidentally, LoadJS does this elegantly because files are loaded asynchronously by default so the load/execution order doesn't matter.

What is the disadvantage of using a pub/sub library?

Let me know your thoughts.

**Comment by wgehner** [comment]
I don't see loading order as an issue because I can put the 'pseudo-bundles` last: `depp.require(['plugin1', 'plugin2', 'xxdone']`. Anyway I've always used loadjs with async: false. 

I don't see how pub/sub would address the requirement.

More importantly, philosophically, an inline
```
<script>
x =5;
</script>
```

is equivalent to

`<script src='/xxfile.js'/>`

where xxfile.js has
`x = 5;`

The inline version is the base case, the external script just a variation of convenience. I would be great if depp dependency management can support both.

**Comment by amorey** [comment]
The difference between the inline version and the external script version is that the browser can control execution time but external script cannot (unless the files are loaded with xhr which has other limitations).

Consider this inline code:
```html
<script src="/file1.js"></script>
<script>x = 5;</script>
<script src="/file2.js"></script>
<script>console.log(x);</script>
```

The browser will download "/file1.js" and "/file2.js" in parallel but execute them in the order defined in the html. JohnnyDepp doesn't have the ability to control execution time so in order to guarantee execution order you would have to do this:
```javascript
depp.define({'plugin1': ['/file1.js'], 'plugin2': ['/file2.js']});

depp.require(['plugin1'], function() {
  x = 5;
  depp.require(['plugin2'], function() {
    console.log(x);
  });
});
```

**Comment by wgehner** [comment]
That seems to me all the more reason why we need a depp.done('xxdone') after x = 5;. 

It sounds like depp.define({'xxdone':[]} might do the job, it just looks more obscure.

**Comment by wgehner** [comment]
Another use case is where we depend on events like Phonegap deviceReady, and need to signal with .done('device'). Or jQuery documentReady with .done('dom'). These are common use cases, and a dependency library might seek to address them.

**Comment by amorey** [comment]
There's some overlap between dependency management and pub/sub but execution order breaks the symmetry.

If you want to add a `depp.done()` method to JohnnyDepp I would fork the code, set `async=true` and then add a `depp.done()` method that calls the `publish()` method internally:
https://github.com/muicss/johnnydepp/blob/master/src/depp.js#L149

Also, if you're only targeting Safari/Chrome/Firefox/Edge you can use html5 modules for dependency management.

**Comment by amorey** [comment]
@wgehner I added a `.done()` method in a way that addresses the concerns I had. The new code is in the [donefn](https://github.com/muicss/johnnydepp/tree/donefn) branch. Here's a version of depp.js with the new code:
https://cdn.rawgit.com/muicss/johnnydepp/donefn/test/assets/johnnydepp/depp.js

Now this will work:
```javascript
depp.define({
  'plugin1': ['/file1.js'],
  'plugin2': ['/file2.js']
});

depp.require(['plugin1', 'xxdone', 'plugin2'], function() {
  console.log(x);  // print 5
});

x = 5;
depp.done('xxdone');
```

As a side effect, the require method will wait to execute until all bundles have been defined. This works for nested bundles too which now must be signaled using a '#' character:
```javascript
// plugin2 depends on plugin1
depp.define({'plugin2': ['#plugin1', '/file2.js']});

// require will execute after all bundles have been defined
depp.require(['plugin2'], function() { /* plugin1 & plugin2 loaded */ });

// defining plugin1 will execute the require() method defined above
depp.define({'plugin1': ['/file1.js']});
```

Please try it out and let me know what you think.

**Comment by wgehner** [comment]
That's awesome! Thank you so much! I just tried it and have a question. Should the following still work or would have to be expressed differently:

```
depp.define({'plugin1': ['/file1.js'], 'plugin2': ['/file2.js']});

depp.require(['plugin1'], function() {
  x = 5;
  depp.require(['plugin2'], function() {
    console.log(x);
  });
});
```
I have a code situation where the x=5 happens but the console.log does not seem to be reached.

**Comment by amorey** [comment]
@wgehner I got your comment via email but I don't see it here. The example you gave should work. Here's a JSFiddle with the working example:
https://jsfiddle.net/muicss/qyjdtc6s/

**Comment by amorey** [comment]
Awesome! It's great to hear it's working well. I'll merge into the master and publish a new version later today.

**Comment by wgehner** [comment]
Perfect, thanks!

**Comment by amorey** [comment]
@wgehner @cekvenich I just pushed a new version that implements the `.done()` method (v0.0.3):
https://www.npmjs.com/package/johnnydepp

Thanks for your help with the new feature! Let me know if you run into any issues with the library.
